core/metacling/src/TClingGenericInfo.h

```diff
@@ -0,0 +1,46 @@
+/// \file TClingGenericInfo.h
+///
+/// \brief The file contains a base class of TCling*Info classes.
+///
+/// \author Vassil Vassilev <vvasilev@cern.ch>
+///
+/// \date March, 2019
+///
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TClingGenericInfo
+#define ROOT_TClingGenericInfo
+
+#include <string>
+
+namespace clang {
+   class Decl;
+}
+
+class TClingGenericInfo {
+protected:
+   const clang::Decl* fDecl = nullptr;
+   std::string fNameCache;
+public:
+   TClingGenericInfo(const clang::Decl* D) : fDecl(D) {}
+   TClingGenericInfo(const TClingGenericInfo& other) {
+      fDecl = other.fDecl;
+      fNameCache = ""; // we must not copy the cache.
```

Comment:
why must we not copy the cache?

Reply:
I still have no idea. Copying it makes hell break loose. I do not quite understand why we have copy ctors and assignement ops all over the place yet. Once I get green light on this I can continue refactoring.

Reply:
FWIW, copying the cache leads to crashes and things like this https://paste.ubuntu.com/p/pZZSBp2HDV/

I suspect something gets copied where it should not have...

Reply:
I suspect the cache is not reset (somehow) when fDecl changes  (later)

Reply:
I think I got it. The iterator should reset the cache because the class could have been in single-decl mode.